pkg/controller/release/release_controller.go

```diff
@@ -283,30 +275,146 @@ func (c *Controller) syncOneReleaseHandler(key string) error {
+	}
+	releases, err := c.releaseLister.Releases(rel.Namespace).ReleasesForApplication(appName)
+	if err != nil {
+		return nil, err
+	}
+	return releases, nil
+}
+
+func (c *Controller) ensureReleaseState(curr *shipper.Release) (*shipper.Release, error) {
+	if curr == nil {
+		// TODO: shippererror
+		return nil, fmt.Errorf("non-empty release expected, nil provided")
+	}
+	releases, err := c.applicationReleases(curr)
+	if err != nil {
+		return nil, err
+	}
+	prev, succ, err := releaseutil.GetSiblingReleases(curr, releases)
+	if err != nil {
+		return nil, err
+	}
+
+	relInfoCurr, err := c.buildReleaseInfo(curr)
+	if err != nil {
+		return nil, err
+	}
+	var relInfoPrev, relInfoSucc *releaseInfo
+	if prev != nil {
+		relInfoPrev, err = c.buildReleaseInfo(prev)
+		if err != nil {
+			return nil, err
+		}
+	}
+	if succ != nil {
+		relInfoSucc, err = c.buildReleaseInfo(succ)
+		if err != nil {
+			return nil, err
 		}
 	}
 
+	//TODO: this flag is deprecated, we keep it here for the sake of gradual
+	//changes. The flag propagates as deep as
+	hasIncumbent := len(releases) > 1
+	executor, err := NewStrategyExecutor(relInfoCurr, relInfoPrev, relInfoSucc, c.recorder, hasIncumbent)
 	if err != nil {
-		return err
+		return nil, err
 	}
 
-	appKey, err := c.getAssociatedApplicationKey(rel)
+	appName, err := releaseutil.ApplicationNameForRelease(curr)
 	if err != nil {
-		return err
+		return nil, err
+	}
+	app, err := c.applicationLister.Applications(curr.Namespace).Get(appName)
+	if err != nil {
+		return nil, err
 	}
 
-	// If everything went fine, scheduling an application key in the
-	// application workqueue.
-	klog.V(4).Infof("Scheduling Application key %q", appKey)
-	c.applicationWorkqueue.Add(appKey)
+	patches, trans, err := executor.Execute()
+	if err != nil {
+		releaseSyncedCond := apputil.NewApplicationCondition(
+			shipper.ApplicationConditionTypeReleaseSynced,
+			corev1.ConditionFalse,
+			conditions.StrategyExecutionFailed,
+			fmt.Sprintf("failed to execute application strategy: %q", err),
+		)
+		diff := apputil.SetApplicationCondition(&app.Status, *releaseSyncedCond)
+		_, err = c.clientset.ShipperV1alpha1().Applications(app.Namespace).Update(app)
+		if err != nil {
+			return nil, shippererrors.NewKubeclientUpdateError(app, err).
+				WithShipperKind("Application")
+		}
+		if !diff.IsEmpty() {
+			c.recorder.Eventf(app, corev1.EventTypeNormal, "ApplicationConditionChanged", diff.String())
+		}
+		return nil, err
+	}
+
+	for _, t := range trans {
+		c.recorder.Eventf(
+			executor.curr.release,
+			corev1.EventTypeNormal,
+			"ReleaseStateTransitioned",
+			"Release %q had its state %q transitioned to %q",
+			shippercontroller.MetaKey(executor.curr.release),
+			t.State,
+			t.New,
+		)
+	}
 
-	klog.V(4).Infof("Done processing Release %q", key)
+	if len(patches) == 0 {
+		klog.V(4).Infof("Strategy verified, nothing to patch")
+		return curr, nil
+	}
+
+	klog.V(4).Infof("Strategy has been executed, applying patches")
```

Comment:
We should say for which release a strategy has been executed, otherwise we have no context. This might be a little bit harder and maybe not strictly necessary, but it would also be useful if we could know which kinds of patches we're about to apply.